processor/metricstransformprocessor/aggregation_helper.go

```diff
@@ -0,0 +1,254 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package metricstransformprocessor
+
+import (
+	"fmt"
+	"math"
+	"math/rand"
+
+	metricspb "github.com/census-instrumentation/opencensus-proto/gen-go/metrics/v1"
+	"github.com/golang/protobuf/ptypes/timestamp"
+)
+
+// timeseriesGroupByLabelValues is a data structure for grouping timeseries that will be aggregated
+type timeseriesGroupByLabelValues struct {
+	// key is a composite of the label values as a single string
+	// keyToTimeseriesMap groups timeseries by the label values
+	timeseries []*metricspb.TimeSeries
+	// keyToLabelValuesMap maps the keys to the actual label values objects
+	labelValues []*metricspb.LabelValue
+}
+
+// getLabelIdxs gets the indices of the labelSet labels' indices in the metric's descriptor's labels field
+// Returns the indices slice and a slice of the actual labels selected by this slice of indices
+func (mtp *metricsTransformProcessor) getLabelIdxs(metric *metricspb.Metric, labelSet map[string]bool) ([]int, []*metricspb.LabelKey) {
+	labelIdxs := make([]int, 0)
+	labels := make([]*metricspb.LabelKey, 0)
+	for idx, label := range metric.MetricDescriptor.LabelKeys {
+		_, ok := labelSet[label.Key]
+		if ok {
+			labelIdxs = append(labelIdxs, idx)
+			labels = append(labels, label)
+		}
+	}
+	return labelIdxs, labels
+}
+
+// aggregateTimeseriesGroups attempts to merge each group of timeseries in keyToTimeseriesMap through the specific calculation indicated by aggrType and dataType to construct the new timeseries with new label values indicated in keyToLabelValuesMap
+// Returns the new list of timeseries
+func (mtp *metricsTransformProcessor) aggregateTimeseriesGroups(keyToTimeseriesMap map[string]*timeseriesGroupByLabelValues, aggrType AggregationType, dataType metricspb.MetricDescriptor_Type) []*metricspb.TimeSeries {
+	newTimeSeriesList := make([]*metricspb.TimeSeries, len(keyToTimeseriesMap))
+	idxCounter := 0
+	for _, element := range keyToTimeseriesMap {
+		timestampToPoints, startTimestamp := mtp.analyzeTimeseriesForAggregation(element.timeseries)
+		newPoints := mtp.aggregatePoints(timestampToPoints, aggrType, dataType)
+		// newSingleTimeSeries is an aggregated timeseries
+		newSingleTimeSeries := &metricspb.TimeSeries{
+			StartTimestamp: startTimestamp,
+			LabelValues:    element.labelValues,
+			Points:         newPoints,
+		}
+		newTimeSeriesList[idxCounter] = newSingleTimeSeries
+		idxCounter++
+	}
+
+	return newTimeSeriesList
+}
+
+// analyzeTimeseriesForAggregation uses timestamp as an indicator to group aggregatable points in the group of timeseries, and determines the final value of the start timestamp for this groups of timeseries.
+// Returns a map that groups points by timestamps and the final start timestamp
+func (mtp *metricsTransformProcessor) analyzeTimeseriesForAggregation(timeseries []*metricspb.TimeSeries) (map[int64][]*metricspb.Point, *timestamp.Timestamp) {
```

Comment:
Yea, the new data model definitely seems less complicated than the current one. However, I don't this this function will actually be removed because the main functionality of this is to group things by label values, which is still needed for the new data model. We probably just have to update this from grouping timeseries to grouping points. I definitely agree with treating each timeseries as one point for now, and later we just have to switch the concept of a timeseries to a point. However, I do have a question about it, for now, if a timeseries has multiple points, should I just take the first point I see and log an error? Or I could do keep as it is for now, and transition this to the new format completely later since now I think about it, the transition shouldn't be hard for this.

Reply:
Hmm I thought the work of aggregating by label values was already done by `groupTimeseriesWithNewValue`?

Also if its aggregating by label values why is it in this file since this file is for functions that are relevant to aggregating by either labels keys or label values? 😄

I'm okay with you keeping it as it is if you like - unfortunate timing unfortunately. I'm looking for anything that can make this processor less complex

Reply:
Side note: `groupByLabelValues` is MUCH better name for `groupTimeseriesWithNewValue` if I've understood that correctly

Reply:
Oh sorry! I meant to say grouping points by timestamp. Yea I am ok with changing this, but what should I do with the timeseries in this version if there are multiple points in one timeseries? Should I just log an error and pick a random one?

Reply:
Yea I think that would be reasonable ... but don't make this change. Probably not worth it since you've already got this working.

Reply:
Btw I think I was partially wrong on this. Good thing you didn't change it 😄

Even though point is squashed into timeseries in the new data structure, points can presumably still have different timestamps, so will still need to be grouped by that.